### PR TITLE
[kubeadm] Adds json struct tags to exposed API types

### LINKED ISSUE #1663: Not all api types have a json struct tag
All api types should have a json struct tag, but BootstrapTokenString does not define any for the fields `ID` and `Secret`.

## Versions
All versions with BootstrapTokenString

I would like to embed the InitConfiguration type into a CRD and use controller-gen v0.2 on that CRD. However, there is a new hard requirement (the requirement has always existed, but it is now an error) that every struct field MUST have a json tag.

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR adds json tags to all exposed kubeadm api types

**Which issue(s) this PR fixes**:
Fixes kubernetes/kubeadm#1663

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by chuckha** [comment]
cc @vincepri

File: cmd/kubeadm/app/apis/kubeadm/bootstraptokenstring.go [hunk 1]
```diff
@@ -35,8 +35,8 @@ import (
 // of view and as an authentication method for the node in the bootstrap phase of
 // "kubeadm join". This token is and should be short-lived
 type BootstrapTokenString struct {
-	ID     string
-	Secret string
+	ID     string `json:"id"`
+	Secret string `json:"secret"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/bootstraptokenstring.go]: internal types don't need the tags.

**Review by neolit123 — COMMENTED** [review]
v1beta1 and v1beta2 are locked at this point, even if this change might not hurt much.

https://github.com/kubernetes/kubeadm/issues/1663

> I would like to embed the InitConfiguration type into a CRD and use controller-gen v0.2 on that CRD. However, there is a new hard requirement (the requirement has always existed, but it is now an error) that every struct field MUST have a json tag.

can we read more about this requirement?

/assign @timothysc @rosti @dixudx

**Comment by chuckha** [comment]
@neolit123 

Here is a slack conversation I had in #kubebuilder https://kubernetes.slack.com/archives/CAR30FCJZ/p1562869078272400

**Comment by chuckha** [comment]
I recognize these types are locked so if this cannot merge I understand. In the mean time, we will be copying these types with this diff into this repo: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm

**Comment by neolit123** [comment]
/hold
for review on the competing PR in the affected repository: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/22

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/bootstraptokenstring.go [hunk 2]
```diff
@@ -31,8 +31,8 @@ import (
 // of view and as an authentication method for the node in the bootstrap phase of
 // "kubeadm join". This token is and should be short-lived
 type BootstrapTokenString struct {
-	ID     string
-	Secret string
+	ID     string `json:"id"`
+	Secret string `json:"secret"`
```
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/bootstraptokenstring.go]: Technically speaking, these fields should not be marshalled as they are. That's why `BootstrapTokenString` implements the `json.Marshaler` interface. Hence, they should simply be omitted (`json:"-"`).
Another possibility is to add NOP tags - `json:"ID"`, `json:"Secret"`.
Although, IDK what are the rules and constraints of the verifier spec, that you are using.

Comment (chuckha) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/bootstraptokenstring.go]: `-` works for me and the verification script I've used. I will update this PR, thanks for the note

**Comment by chuckha** [comment]
@neolit123 Does this still need the hold? That PR merged

**Review by rosti — COMMENTED** [review]
Thanks @chuckha !
Let's squash the commits. I am OK with this being merged as is. It doesn't change anything, except it plugs kubebuilder's verifier.
@fabriziopandini @neolit123 WDYT?

**Comment by chuckha** [comment]
OK! fixed the commits (heheh, meant to do that before, thanks for catching)

**Comment by neolit123** [comment]
i was -1 for making the change because:
- we should consider the types locked at this point
- the consuming repository has to fork the types regardless for now and the workaround can be applied there.

but if @rosti and @fabriziopandini are OK with this k/k change go ahead.

**Comment by fabriziopandini** [comment]
> It doesn't change anything, except it plugs kubebuilder's verifier. 

+1 from my side (BootstrapTokenString has its own Marshal/UnMarshal, so this change should not affect users

**Comment by rosti** [comment]
Thank you all for the feedback!
I'll let this in.
/approve
/lgtm
/hold cancel

**Pull request merged**